traitlets/traitlets.py

```diff
@@ -447,22 +447,25 @@ def __get__(self, obj, cls=None):
                 return value
 
     def __set__(self, obj, value):
-        new_value = self._validate(obj, value)
-        try:
-            old_value = obj._trait_values[self.name]
-        except KeyError:
-            old_value = self.default_value
+        if self._metadata.get('read_only') and obj._read_only_enabled:
```

Comment:
I'm also not terribly pleased that this requires an extra check every time any trait gets set. I know it's not much overhead, it just seems like bad design.